---
title: "The search for Schizosaccharomyces yeast in environmental metatranscriptomes"
author: "R. Shraim & B. Nieuwenhuis"
date: "`r Sys.Date()`"
output: html_document
---

```{r input-data, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
#set directory for blast results, NC ids of schizo, processed SRAs, SRA metadata, common names

#setwd(" ") 


library(tidyverse)
library(scales)
library(ggpubr)
library(RColorBrewer)
library(knitr)
library(ape)
library(sp)
library(IRanges)
library(bookdown)
library(VennDiagram)
library(ggrepel)
library(png)
library(viridis)

myviridis <- c("Schizosaccharomyces pombe" = "#482677FF", "Schizosaccharomyces cryophilus" = "#2D708EFF", "Schizosaccharomyces octosporus" = "#55C667FF","Schizosaccharomyces japonicus"= "#FDE725FF",
               "S. pombe" = "#482677FF", "S. cryophilus" = "#2D708EFF", "S. octosporus" = "#55C667FF","S. japonicus"= "#FDE725FF")

#read NC (chromosome) ids for the schizosaccharomyces species
schizosaccharomyces_NC <- read_table2("schizosaccharomyces_NC.txt", 
                          col_names = F) %>% 
                          unite(col = "species", c("X2", "X3"), sep=" ")

#read list of processed Runs (to reduce size of SRA metadata table)
mybatches <- read_table("allbatches_accessions.txt", col_names = F)
colnames(mybatches) <- c("Run")

#read metadata from SRA table and select info of interest  & processed Runs
sra <- read_csv("allbatches_SraRunInfo_illumina_metatranscript.csv", col_names = T) %>% 
        select("BioProject", "Run", "ScientificName", "spots") %>% distinct()

#filter for processed runs only
sra <- filter(sra, Run %in% mybatches$Run)

#read group/common name info (includes all processed BioProjects/Runs)
env <- read_csv("metatranscriptome_env_groups.csv") 

#join metadata SRA table and group/name info
sra <- left_join(sra, env, by=c("BioProject", "ScientificName"))

#read data from BLAST for reads mapped to schizosaccharomyces and other fungi (filtered for pident > 98, cover > 0.9)
blast <- read_table2("all_hits_full_blast.db", 
                     col_names = c("qseqid", "sseqid","pident","qlen","length","qstart",
                                   "qend","sstart","send","evalue", "cover"))


```


#### Positive control: RNA-seq of S. pombe figure 2
```{r pombevalidation, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}
#SRR3745342	marine metagenome
#SRR7291883	pombe rna

#read in blast table with 0.01 percent reads
blast001 <- read_table2("rnapombe/fake001.txt", col_names = F)  
colnames(blast001) <- c("qseqid","sseqid","pident","qlen","length","qstart","qend","sstart","send","evalue")
blast001$cover <- blast001$length/blast001$qlen

#select top hit by cover then pident then evalue then select only reads that mapped to schizo genomes (not other fungi)
blast001 <- group_by(blast001, qseqid) %>% top_n(1, cover) %>% top_n(1, pident) %>% top_n(-1, evalue) %>% 
  ungroup() %>% 
  filter(sseqid %in% schizosaccharomyces_NC$X1)

#join with schizosaccharomyces_NC to get species name for sseqid and keep only one hit for each read
results001 <- inner_join(blast001, schizosaccharomyces_NC, by = c("sseqid" = "X1")) %>% 
  select(qseqid, pident, evalue, cover, species) %>% 
  distinct()

#change read name to accession and match accession number to either marine metagenome or S. pombe
results001$qseqid <- str_replace(results001$qseqid, "\\..*", "")
results001$ScientificName <- ifelse(results001$qseqid == "SRR3745342", "marine metagenome", "S.pombe")

#filter by pident and cover and count number of reads that mapped to each schizo species
results001 <- filter(results001, pident > 98, cover > 0.9) %>% 
  group_by(species, ScientificName) %>% summarize( n = n())
results001$total_added <- 600
results001$percent <- round(results001$n/results001$total_added * 100, digits = 2)

blast01 <- read_table2("rnapombe/fake01.txt", col_names = F)  
colnames(blast01) <- c("qseqid","sseqid","pident","qlen","length","qstart","qend","sstart","send","evalue")
blast01$cover <- blast01$length/blast01$qlen
blast01 <- group_by(blast01, qseqid) %>% top_n(1, cover) %>% top_n(1, pident) %>% top_n(-1, evalue) %>% ungroup() %>%
  filter(sseqid %in% schizosaccharomyces_NC$X1)
results01 <- inner_join(blast01, schizosaccharomyces_NC, by = c("sseqid" = "X1"), copy=F) %>%
  select(qseqid, pident, evalue, cover, species) %>%
  distinct()
results01$qseqid <- str_replace(results01$qseqid, "\\..*", "")
results01$ScientificName <- ifelse(results01$qseqid == "SRR3745342", "marine metagenome", "S.pombe")
results01 <- results01 %>% filter(pident > 98, cover > 0.9) %>% 
  group_by(species, ScientificName) %>% summarize( n = n())
results01$total_added <- 6300
results01$percent <- round(results01$n/results01$total_added * 100, digits = 2)


blast1 <- read_table2("rnapombe/fake1.txt", col_names = F)  
colnames(blast1) <- c("qseqid","sseqid","pident","qlen","length","qstart","qend","sstart","send","evalue")
blast1$cover <- blast1$length/blast1$qlen
blast1 <- group_by(blast1, qseqid) %>% top_n(1, cover) %>% top_n(1, pident) %>% top_n(-1, evalue) %>% ungroup() %>%
  filter(sseqid %in% schizosaccharomyces_NC$X1)
results1 <- inner_join(blast1, schizosaccharomyces_NC, by = c("sseqid" = "X1"), copy=F) %>%
  select(qseqid, pident, evalue, cover, species) %>%
  distinct()
results1$qseqid <- str_replace(results1$qseqid, "\\..*", "")
results1$ScientificName <- ifelse(results1$qseqid == "SRR3745342", "marine metagenome", "S.pombe")
results1 <- results1 %>% filter(pident > 98, cover > 0.9) %>% 
  group_by(species, ScientificName) %>% summarize( n = n())
results1$total_added <- 64000
results1$percent <- round(results1$n/results1$total_added * 100, digits=2)


blast5 <- read_table2("rnapombe/fake5.txt", col_names = F)  
colnames(blast5) <- c("qseqid","sseqid","pident","qlen","length","qstart","qend","sstart","send","evalue")
blast5$cover <- blast5$length/blast5$qlen
blast5 <- group_by(blast5, qseqid) %>% top_n(1, cover) %>% top_n(1, pident) %>% top_n(-1, evalue) %>% ungroup() %>%
  filter(sseqid %in% schizosaccharomyces_NC$X1)
results5 <- inner_join(blast5, schizosaccharomyces_NC, by = c("sseqid" = "X1"), copy=F) %>%
  select(qseqid, pident, evalue, cover, species) %>%
  distinct()
results5$qseqid <- str_replace(results5$qseqid, "\\..*", "")
results5$ScientificName <- ifelse(results5$qseqid == "SRR3745342", "marine metagenome", "S.pombe")
results5 <- results5 %>% filter(pident > 98, cover > 0.9) %>% 
  group_by(species, ScientificName) %>% summarize( n = n())
results5$total_added <- 334000
results5$percent <- round(results5$n/results5$total_added * 100, digits = 2)

fakepombe <- dplyr::bind_rows(list("0.01% reads"=results001, "0.1% reads"=results01, "1% reads"=results1, "5% reads"=results5), .id = 'source')

fpombeplot <- ggplot(fakepombe, aes(x = species, y=percent, fill=species)) + 
    geom_bar(position = "dodge", stat = "identity") +
  theme_bw() +  
  theme(text=element_text(size=12),
          axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
    scale_fill_manual(values = myviridis) + 
  facet_wrap(~source, ncol=4) +
 geom_text(aes(label=percent), vjust=-0.2) +
  ylab("percentage of S. pombe reads recovered") +
  xlab("")

```

#### Identical mapping: venn diagram and overlap groups figure 2

```{r overlap, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}

#Overlap between Schizosaccharomyces spp. and the other fungi group in the BLAST database by accession and by read

schizo <- c("Schizosaccharomyces octosporus", "Schizosaccharomyces japonicus",
            "Schizosaccharomyces pombe", "Schizosaccharomyces cryophilus")

#read in names of fungi in blast fungal db
fungi <- read_csv("supplementary_table_S2_fungi.csv")

fungi_names <- unique(fungi$other)
fungi_names

#join blast results with the fungi names (sseqid in blast results is by contig)
allblast <- left_join(blast, fungi, by=c("sseqid"="seq") )

#select reads and accessions and which fungi they map to 
myreads <- select(allblast, qseqid, other) %>% distinct()
myreads$run <- str_replace(myreads$qseqid, "\\..*", "")
myreads


###
#adding in the reads that mapped only to non-schizo fungi in the first BLAST alignment (skip if doing one BLAST alignment)
originalonlyother <- read_table2("venn_diagram/venn_onlyotherfungi_reads_originaldb.txt")
###

#separate reads to two lists, reads that map to schizo and reads that map to other fungi
set_schizo <- filter(myreads, other %in% schizo) %>% select(qseqid) %>% unlist()
set_other <- filter(myreads, !other %in% schizo) %>% select(qseqid) %>% bind_rows(originalonlyother) %>% unlist()

#for accessions
#myreads <- select(myreads, run, fungus) %>% distinct()
#set_schizo <- filter(myreads, fungus %in% schizo) %>% select(run) %>% unlist()
#set_other <- filter(myreads, !fungus %in% schizo) %>% select(run) %>% unlist()

#create a venn diagram that shows overlap between reads mapping to schizo and reads mapping to other fungi
plt_venn <- venn.diagram(
  x = list(set_schizo, set_other),
  category.names = c("Schizosaccharomyces", "Other fungi"),
  filename = NULL,
  col=c("#fd7e4d", '#000000'),
  fill = c(alpha("#fd7e4d",0.8), alpha('#000000',0.8)),
  cex = 1,
  fontfamily = "sans",
  cat.cex = 1.1,
  #cat.default.pos = "outer",
  cat.pos = c(0,0),
  cat.dist = c(-0.1, -0.1),
  cat.fontfamily = "sans"
  #cat.col = c("#008b8a", '#000000')
)
grid::grid.newpage()
grid::grid.draw(plt_venn)


#looking into the details of the overlapping reads/accessions
#make tibbles of reads that map to only schizo, only other fungi, or both
onlyschizo <- as_tibble(setdiff(set_schizo, set_other)) %>% distinct()
onlyschizo <- separate(onlyschizo, value, c("Run", "Read"), remove = F )

onlyother <- as_tibble(setdiff(set_other, set_schizo)) %>% distinct()
onlyother <- separate(onlyother, value, c("Run", "Read"), remove = F )

overlap <- as_tibble(intersect(set_schizo, set_other)) 
overlap <- separate(overlap, value, c("Run", "Read"), remove = F )
overlap <- left_join(overlap, sra)

#count number of accessions and number of reads in each environmental group where reads overlap and in the reads that don't overlap
overlapcount <- overlap %>% group_by(Group) %>% summarize(n_reads=n())
overlap_acc <- select(overlap, Group, Run) %>% distinct() %>% group_by(Group) %>% summarise(n_accessions = n())
overlapcount <- left_join(overlapcount, overlap_acc)

df <- tibble("Group"=c("only schizosaccharomyces", "only other fungi"), 
           "n_reads"=c(length(unique(onlyschizo$value)), length(unique(onlyother$value))), 
           "n_accessions"=c(length(unique(onlyschizo$Run)), length(unique(onlyother$Run)))
           )
overlapcount <- bind_rows(overlapcount, df)
overlapcount

#plot the number of reads and accessions in the overlapping groups with n accessions on y axis and size of dot corresponding to n reads
overlapplot <- ggplot(overlapcount, aes(x=reorder(Group, -n_reads), y=n_accessions, size = n_reads)) + 
  geom_point(alpha =0.5) + 
  theme_bw() +
  theme(axis.text = element_text(size=14)) +
 # theme(legend.position = "none") + 
  scale_size(range = c(1, 20), name="number of reads") +
  xlab("") +
  ylab("total number of accessions per Group") +
  coord_flip()

figure2 <- ggarrange(fpombeplot, 
  ggarrange(as_ggplot(grobTree(plt_venn)), overlapplot, ncol=2, nrow=1, widths = c(1,2),  labels = c("B", "C")),
  nrow=2, labels = "A")

figure2

ggsave(figure2, filename = "figure2.pdf", width = 12, height = 9)

```

#### correlation test

```{r}

#correlation test
#test correlation between number of positive fungi hits and number of total spots in the accession
spotcor <- filter(myreads, other %in% schizo) %>% group_by(run) %>% summarise(n=n()) %>% left_join(select(sra, Run, spots), by=c("run"="Run"))

#data not normally distributed
cor.test(spotcor$n, spotcor$spots, method = "kendall")


```


#### proportion of hits per environment: graphical abstract

```{r proportion, echo=FALSE, message=FALSE,warning=FALSE, cache=FALSE}

results <- filter(allblast, qseqid %in% onlyschizo$value)
colnames(results) <- c("qseqid", "sseqid","pident","qlen","length", "qstart","qend","sstart","send","evalue", "cover", "species")

#keep only unique read hits per schizo species
schizo_hits <- select(results, qseqid, species) %>% distinct()
schizo_hits

schizo_hits <- filter(schizo_hits, !qseqid %in% overlap$value)

#add run and read ids from qseqid
schizo_hits <- separate(schizo_hits, qseqid, c("Run", "Read"), remove = F )
schizo_hits

#count number of unique mapped reads per schizo species per accession
uniq_schizo_hits <- select(schizo_hits, Run, species)  %>% group_by(Run, species) %>% dplyr::summarize(n_reads=n()) 
uniq_schizo_hits

#join with the SRA info
uniq_schizo_hits <- left_join(uniq_schizo_hits, sra, by="Run")
#uniq_schizo_hits <- filter(uniq_schizo_hits, Group != "human") #discard human environments
uniq_schizo_hits

#join list of processed accessions with the SRA info
mybatches <- left_join(mybatches, sra, by=c("Run"="Run")) 
mybatches

#create df with mapped/not mapped info
#mark unique accessions with hits in them as "present" 
withhits <- tibble(Run = unique(uniq_schizo_hits$Run), 
                  mapped = rep("present", length(unique(uniq_schizo_hits$Run))))
#join accessions that are marked "present" with all processed accessions and mark the rest "absent"
mybatches <- left_join(mybatches, withhits, by=c("Run"="Run"))
mybatches <- mutate(mybatches, mapped = replace_na(mapped, "absent"))
mybatches


#### graphical abstract
#percentage of ACCESSIONS in which schizo was present of the tested accessions/group

#count number of accessions with or without hits in each environmental group
batchplot_acc <- mybatches %>% group_by(Group, mapped) %>% dplyr::summarize(n_acc=n())

grabstract <- batchplot_acc %>% 
  ggplot( aes(x = Group, y=n_acc, fill=mapped)) + 
  geom_bar(position = "fill", stat = "identity") + 
    theme_classic() +
  theme(axis.title.x = element_text(size=14),
    axis.text.x = element_text(size=14),
        axis.text.y = element_text(size=14),
    axis.title.y = element_blank(),
        legend.position="top",
        legend.title = element_blank(),
        legend.text = element_text(size=14)) +
 scale_fill_viridis(discrete = T) +
    scale_y_continuous(breaks = seq(0, 1, by=0.5)) +
  ylab("proportion of SRA accessions with Schizosaccharomyces") +
    coord_flip() 

grabstract


#percentage of BIOPROJECTS in which schizo was present of the tested bp/group
#count number of bioprojects with or without hits in each environmental group
batchplot_bp <- distinct(select(mybatches, BioProject, Group, mapped)) %>% 
                    group_by(Group, mapped) %>% dplyr::summarize(n_bp=n())

#to show the total number of bioprojects per environmental group
totals <- batchplot_bp %>% group_by(Group) %>% summarize(total = sum(n_bp))

#change fill to stack for count instead of proportion
bpprop <- batchplot_bp %>% 
  ggplot( aes(x = Group, y=n_bp, fill=mapped)) + 
  geom_bar(position = "fill", stat = "identity") + 
  theme(axis.title.x=element_blank(),
        axis.text.x = element_text(angle=45, vjust=1, size = 10),
        text=element_text(size=10),
        legend.position="top",
        legend.title = element_blank()) +
  theme_classic() +
  ylab("proportion of BioProjects in which Schizosaccharomyces spp. was present") +
  xlab(" ") +
  coord_flip()


```

#### range of environments processed: figure 1

```{r rangeofgroups}
#with the pipeline figure, to show range of environments and number of reads we processed
env_range <- select(mybatches, Group, ScientificName, spots) %>% distinct()

#spots is the number of reads
env_count <- group_by(env_range, Group) %>% summarise(n_spots = sum(spots)) 

n_scientificname <- select(env_range, Group, ScientificName) %>% distinct() %>% 
  group_by(Group) %>% summarise(n_scientificname=n()) 

env_count <- left_join(env_count, n_scientificname)

env_count

envplot <- ggplot(env_count, aes(x=(reorder(Group, -n_scientificname)), y=n_spots, size = n_scientificname)) + 
  geom_point(alpha =0.5) + 
  theme_bw() +
  theme(axis.text = element_text(size=10),
        legend.text = element_text(size=10),
        legend.title = element_text(size=10),
        axis.title.x = element_text(size=11)) +
 # theme(legend.position = "none") + 
  scale_size(range = c(1, 10), name="number of ScientificName") +
  xlab("") +
  ylab("total number of reads per Group") +
  coord_flip() 

imgpipeline <- readPNG("pipeline.png")

imgp <- ggplot() + background_image(imgpipeline)

figure1 <- ggarrange(imgp, envplot, nrow = 2, heights = c(5,3), labels = "AUTO")

figure1

ggsave(figure1, filename = "figure1.pdf", width = 9, height = 12)
```


#### normalized counts of read/acc/bp hits per group: figure 3

```{r countenvironments, echo=FALSE, message=FALSE,warning=FALSE, cache=FALSE}


#at READ level
#count number of reads per species per group
readcounts <- left_join(schizo_hits, sra) %>% group_by(Group, species) %>% dplyr::summarize(n_reads=n())
readcounts

#total number o fspots per group to normalize by
norm_spots <- mybatches %>% select(Group, spots) %>% group_by(Group) %>% summarize(n=sum(spots))
norm_spots

readcounts <- left_join(readcounts, norm_spots)
readcounts$n_norm <- readcounts$n_reads / readcounts$n
readcounts$n_label <- round(readcounts$n/100000000, digits=2)
readcounts$group_label <- paste0(readcounts$Group, " (", readcounts$n_label, ")")


#at ACCESSION level
#total number of accessions processed per group
norm_acc <- select(mybatches, Run, Group) %>% group_by(Group) %>% summarize(n=n())
norm_acc

#count number of accessions per species per group
acc_env_sp <- uniq_schizo_hits %>% group_by(Group, species) %>% dplyr::summarize(n_acc=n()) %>% 
  arrange(by=n_acc)

acc_env_sp <- left_join(acc_env_sp, norm_acc)
acc_env_sp$n_norm <- acc_env_sp$n_acc / acc_env_sp$n
acc_env_sp$group_label <- paste0(acc_env_sp$Group, " (", acc_env_sp$n, ")")


#similarly, at BIOPROJECT level
#number of bioprojects processed per group
norm_bp <- select(mybatches, BioProject, Group) %>% distinct() %>% group_by(Group) %>% summarize(n=n())

bp_env_sp <- ungroup(uniq_schizo_hits) %>% select(species, BioProject, Group) %>% distinct()
bp_env_sp <- bp_env_sp %>% group_by(Group, species) %>% dplyr::summarize(n_bp=n())

bp_env_sp <- left_join(bp_env_sp, norm_bp)
bp_env_sp$n_norm <- bp_env_sp$n_bp / bp_env_sp$n
bp_env_sp$group_label <- paste0(bp_env_sp$Group, " (", bp_env_sp$n, ")")

#plot number of hit reads per group normalized by total number of reads in group
readplot <- readcounts %>% 
  ggplot( aes(x = reorder(group_label, n_norm), y=n_norm, fill=species)) + 
  geom_bar(position = "dodge", stat = "identity", alpha=0.8) +
  theme_bw() +
  theme(axis.text = element_text(size = 10),
        text=element_text(size=10)) +
  ylab("hit reads/spots") +
  xlab(" ") +
  scale_colour_manual(values=myviridis, aesthetics = c("colour", "fill"), labels = c("S. cryophilus", "S. japonicus", "S. octosporus", "S. pombe")) +
  coord_flip() 

accplot <- acc_env_sp %>% 
  ggplot( aes(x = reorder(group_label, n_norm), y=n_norm, fill=species)) + 
  geom_bar(position = "dodge", stat = "identity", alpha=0.8) +
  theme_bw() +
  theme(axis.text = element_text(size = 10),
        text=element_text(size=10)) +
  ylab("hit accessions/total accessions") +
  xlab(" ") +
  scale_colour_manual(values=myviridis, aesthetics = c("colour", "fill"), labels = c("S. cryophilus", "S. japonicus", "S. octosporus", "S. pombe")) + 
  coord_flip() 

bpplot <- bp_env_sp %>% 
  ggplot( aes(x = reorder(group_label, n_norm), y=n_norm, fill=species)) + 
  geom_bar(position = "dodge", stat = "identity", alpha=0.8) +
  theme_bw() +
  theme(axis.text = element_text(size = 10),
        text=element_text(size=10)) +
  ylab("hit BioProjects/total BioProjects") +
  xlab(" ") +
  scale_colour_manual(values=myviridis, aesthetics = c("colour", "fill"), labels = c("S. cryophilus", "S. japonicus", "S. octosporus", "S. pombe")) +
  coord_flip()

figure3 <- ggarrange(accplot, bpplot, ncol=2, nrow=1,  labels = "AUTO",common.legend = TRUE, legend="top")

figure3 

ggsave(figure3, filename = "figure3.pdf", width = 13, height = 9)
```

#### Hit genomic position: figure 4

```{r position, echo=FALSE, message=FALSE, warning=FALSE, cache=FALSE}

#select start and end for genomic position from blast results
gresults <- filter(schizo_hits, species == "Schizosaccharomyces pombe") %>% left_join(results)
gresults <- distinct(gresults) 

#### Schizosaccharomyces pombe
pombe <- left_join(gresults, sra)
#pombe <- filter(pombe, Group != "human")
#replace NCs by chromosome name and add length of previous chromosome
pombe$sseqid <- str_replace(pombe$sseqid, "NC_003424.3", "chr_I")
pombe$sseqid <- str_replace(pombe$sseqid, "NC_003423.3", "chr_II")
pombe$sseqid <- str_replace(pombe$sseqid, "NC_003421.2", "chr_III")
pombe$sseqid <- str_replace(pombe$sseqid, "NC_001326.1", "mit")
pombe$genome <- pombe$sstart
pombe$genome[pombe$sseqid == "chr_II"] <- pombe$genome[pombe$sseqid == "chr_II"] + 5579133
pombe$genome[pombe$sseqid == "chr_III"] <- pombe$genome[pombe$sseqid == "chr_III"] + 10118937
pombe$genome[pombe$sseqid == "mit"] <- pombe$genome[pombe$sseqid == "mit"] + 12571820
pombe$qseqid <- as.factor(pombe$qseqid) 

chr <- c("chr_I", "chr_II", "chr_III")
winSize = 5000
pombechr <- filter(pombe, sseqid %in% chr) %>% mutate( index = ceiling(  sstart / winSize )* winSize ) %>%
  mutate( index_chr = ceiling(  genome / winSize )* winSize ) %>%
  arrange(genome) 

pombe_genome_window <- pombechr %>% group_by(Group, sseqid, index) %>% dplyr::summarize(n_reads=n()) 

mwinSize = 100
pombemit <- filter(pombe, sseqid=="mit" ) %>% mutate( index = ceiling(  sstart / mwinSize )* mwinSize ) %>%
  mutate( index_chr = ceiling(  genome / mwinSize )* mwinSize ) %>%
  arrange(genome) 

pombe_mit_window <- pombemit %>% group_by(Group, sseqid, index) %>% dplyr::summarize(n_reads=n()) 


pombechrplot <- droplevels(pombe_genome_window) %>% 
ggplot(aes(x = index, y = Group, fill = log10(n_reads ) ) ) +
  geom_tile() +
  theme_bw() +
  scale_fill_viridis( limits = c(0, ceiling(max( log10( pombe_genome_window$n_reads ) ) ) ) )+
  theme(axis.text.x = element_text( vjust=1, size=10),
         axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 11)) +
  guides(fill=guide_legend(ncol=1)) +
  facet_wrap( sseqid  ~ . , scales = "free", ncol = 1, strip.position = "right") 


pombemitplot <- droplevels(pombe_mit_window) %>% 
ggplot(aes(x = index, y = Group, fill = log10(n_reads ) ) ) +
  geom_tile() +
  theme_bw() +
  scale_fill_viridis( limits = c(0, ceiling(max( log10( pombe_mit_window$n_reads ) ) ) ) )+
  theme(axis.text.x = element_text( vjust=1, size=10),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 11)) +
  guides(fill=guide_legend(ncol=1)) +
  facet_wrap( sseqid  ~ . , scales = "free", ncol = 1, strip.position = "right") +
  ylab("S. pombe chromosomes I-III (winsize 5000), mitochondria (winsize 100)") 
   

figure4 <- ggarrange(pombechrplot, pombemitplot, nrow=2, heights = c(3,1), common.legend = T, legend = "right")

figure4

ggsave("figure4.pdf", figure4)

```

#### supplementary figure

```{r}

winSize = 75000
TMP <- pombe %>% 
  mutate( index = ceiling(  sstart / winSize )* winSize ) %>%
  mutate( index_chr = ceiling(  genome / winSize )* winSize ) %>%
  group_by(Group, BioProject, index_chr) %>% 
  dplyr::summarize(n_reads=n()) 
TMP %>%
  ggplot(aes(x = index_chr, y = BioProject, fill = log10(n_reads ) ) ) +
  geom_tile() +
  scale_fill_gradientn(colours=c('deepskyblue', 'red'), 
                       na.value = "green", limits = c(min(log10(TMP$n_reads ) ),max(log10(TMP$n_reads )  ))) +  
  xlim(0,max(pombe$genome)) +
  theme_bw() +
  theme(#strip.background = 'blue',
        #strip.text.y = element_blank(),
        axis.text.x = element_text(  size=12 ),
        legend.title = element_blank(), 
        legend.position = "right",
        axis.title.y = element_blank(),
        axis.text.y = element_text(size = 5, angle = 45 )) +
  guides(fill=guide_legend(ncol=1)) +
  geom_vline( xintercept = c(5579133, 10118937, 12571820))+
  #scale_color_manual(values=myColors) +
  facet_wrap(.   ~  Group , scales = "free", ncol = 5, strip.position = "right") +
  xlab("S. pombe genome") +
  ylab(" ") 



```

```{r transcriptcounts, echo=FALSE,  message=FALSE,warning=FALSE, cache=FALSE}

#use the pombe gengff and IRanges pkg to count number of reads per type of transcripts

cresults <- select(results, qseqid, species, sstart, send, sseqid)
cresults <- distinct(cresults) 

cpombe <- filter(cresults, species=="Schizosaccharomyces pombe")
cpombe$sseqid <- str_replace(cpombe$sseqid, "NC_003424.3", "I")
cpombe$sseqid <- str_replace(cpombe$sseqid, "NC_003423.3", "II")
cpombe$sseqid <- str_replace(cpombe$sseqid, "NC_003421.2", "III")
cpombe$sseqid <- str_replace(cpombe$sseqid, "NC_001326.1", "mitochondrial")

chr <- c("I", "II", "III", "mitochondrial")

tmp <- c("gene", "gene_group", "CDS")

gengff <- as_tibble(read.gff("Schizosaccharomyces_pombe_all_chromosomes.gff3")) %>% 
  filter(!type %in% tmp) %>%
  separate(attributes, sep ="=|;", into = c("id", "gene", "par"), remove = F) %>% 
  #filter(seqid %in% chr) %>% 
  droplevels()

#for IRanges, send cannot be smaller than sstart
#when transformation applied, some rows become duplicated
#use 'gene' instead of 'attributes' to get read count by gene
cpombe <- as_tibble(transform(cpombe, send = ifelse(sstart > send, sstart, send), 
                              sstart = ifelse(send < sstart, send, sstart))) %>% distinct()
rangesP <- split(IRanges(cpombe$sstart, cpombe$send, names = cpombe$qseqid), cpombe$sseqid)
rangesG <- split(IRanges(gengff$start, gengff$end, names = gengff$attributes), gengff$seqid)


#queryHits is index in rangesP and subjectHits is index in rangesG
hits <- findOverlaps(rangesP, rangesG, type = "any") #query, subject, type

chrI <- as_tibble(hits$I) %>% mutate(sseqid=rep("I", length(hits$I)))
chrI$qseqid <- names(rangesP$I[chrI$queryHits])
chrI$attributes <- names(rangesG$I[chrI$subjectHits]) #gene
chrI$sstart <- rangesP$I@start[chrI$queryHits] 
chrI$send <- rangesP$I@start[chrI$queryHits] + rangesP$I@width[chrI$queryHits] -1
chrI$CDSwidth <- rangesG$I@width[chrI$subjectHits]

chrII <- as_tibble(hits$II)%>% mutate(sseqid=rep("II", length(hits$II)))
chrII$qseqid <- names(rangesP$II[chrII$queryHits]) 
chrII$attributes <- names(rangesG$II[chrII$subjectHits]) #gene
chrII$sstart <- rangesP$II@start[chrII$queryHits] 
chrII$send <- rangesP$II@start[chrII$queryHits] + rangesP$II@width[chrII$queryHits] -1
chrII$CDSwidth <- rangesG$II@width[chrII$subjectHits]

chrIII <- as_tibble(hits$III) %>% mutate(sseqid=rep("III", length(hits$III)))
chrIII$qseqid <- names(rangesP$III[chrIII$queryHits]) 
chrIII$attributes <- names(rangesG$III[chrIII$subjectHits]) #gene
chrIII$sstart <- rangesP$III@start[chrIII$queryHits] 
chrIII$send <- rangesP$III@start[chrIII$queryHits] + rangesP$III@width[chrIII$queryHits] -1
chrIII$CDSwidth <- rangesG$III@width[chrIII$subjectHits]

mit <- as_tibble(hits$mitochondrial) %>% mutate(sseqid=rep("mitochondrial", length(hits$mitochondrial)))
mit$qseqid <- names(rangesP$mitochondrial[mit$queryHits, ]) 
mit$attributes <- names(rangesG$mitochondrial[mit$subjectHits, ]) #gene
mit$sstart <- rangesP$mitochondrial@start[mit$queryHits] 
mit$send <- rangesP$mitochondrial@start[mit$queryHits] + rangesP$mitochondrial@width[mit$queryHits] -1
mit$CDSwidth <- rangesG$mitochondrial@width[mit$subjectHits]

#use 'gene' instead of 'attributes' to get read count by gene
allhits <- bind_rows(chrI, chrII, chrIII, mit) %>% 
  select("sseqid","qseqid","sstart", "send", "attributes", "CDSwidth") %>% 
  distinct()

pomgen <- left_join(cpombe, allhits, by=c("sseqid", "qseqid", "sstart", "send")) 
#some reads overlap with multiple CDS since some CDS overlap in the genome

pomgen$qseqid <- str_replace(pomgen$qseqid, "\\..*", "") 
pomgen <- left_join(pomgen, sra, by=c("qseqid"="Run"))

pomgen <- left_join(pomgen, gengff, by="attributes")
#pomgen$gene <- str_replace(pomgen$gene, ":exon.*", "") 

#### count type of transcript
count <- pomgen %>% group_by(type) %>% summarize(n=n()) %>% arrange(desc(n))
knitr::kable(count[1:10,], col.names = c("CDS type", "N reads"),
             caption = "Number of reads per type of coding sequence in S. pombe (showing only top ten types)")



#### or count by gene
#count top genes but normalize by length of the cds
#count <- pomgen %>% group_by(gene) %>% summarize(n=n()) 

#some genes have "multiple widths" because they are divided to exons
#sum exon length per gene
#width <- distinct(select(pomgen, gene, CDSwidth)) %>% group_by(gene) %>% summarize(CDSwidth=sum(CDSwidth))

#normcount <- inner_join(count, width, by = "gene") %>% na.omit()
#normcount$norn <- normcount$n/normcount$CDSwidth
#normcount <- arrange(normcount, desc(norn))

#excluding rRNA genes
#nonrna <- filter(normcount, str_detect(gene, "SPRRNA", negate=T)) %>% arrange(desc(norn))

#knitr::kable( list( normcount[1:5, ], nonrna[1:5, ]), col.names = c("transcript", "n", "transcript length (bp)", "normalized n"),  caption = "Top five CDS transcripts that S. pombe hits fall within, normalized to the transcript length. All of these transcripts are rRNAs. When rRNA is excluded, some of the top transcripts include SPAC4F10.13c (mpd2) a GYF domain protein-coding gene and antisense RNA, SPNCRNA.957. The total of unique reads that mapped to S. pombe was 107,393.", booktabs = TRUE )

```
